### PR TITLE
dcos-419 add support for SSL

### PR DESCRIPTION
**Pull request by tamarrow**
first pass

### CODE REVIEW

**Comment by tamarrow** [comment]
fyi ran into this (since I'm running python 2.7.6): http://stackoverflow.com/questions/29099404/ssl-insecureplatform-error-when-using-requests-package. Will probably want to document it

**Comment by jsancio** [comment]
@tamarrow Do you want me to review this? Tests are not passing.

**Comment by tamarrow** [comment]
@jsancio thats cause ssl_verify is set to true, ill change it to false

**Comment by jsancio** [comment]
Also we are missing tests for the positive and negative cases. Missing files, invalid ssl, etc.

**Comment by tamarrow** [comment]
@jsancio added tests

**Comment by jsancio** [comment]
@tamarrow ping

**Comment by tamarrow** [comment]
@jsancio rebased

File: cli/dcoscli/main.py [hunk 1]
```diff
@@ -79,6 +85,10 @@ def _main():
        not auth.check_if_user_authenticated():
         auth.force_auth()
 
+    #  TODO (tamar): support reading tuple:
+    #  (/path/to/cert, /path/to/private/key)
```
Comment (jsancio) [inline cli/dcoscli/main.py]: What is this TODO about? Do we still need to implement this?

File: cli/tests/integrations/test_ssl.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from dcos import http, util
+from dcos.errors import DCOSException
+
+import pytest
+from six.moves import urllib
+
+DCOS_URL = util.get_config()['core.dcos_url'].replace('http://', 'https://')
+DCOS_URL = urllib.parse.urljoin(DCOS_URL, 'marathon/')
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Not a fan of this coding style. Why not just?
```
DCOS_URL = urllib.parse.urljoin(
    util.get_config()['core.dcos_url'].replace('http://', 'https://'),
    'marathon/')
```

File: cli/tests/integrations/test_ssl.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+
+def test_dont_verify_ssl():
+    resp = http.request("get", DCOS_URL, verify=False)
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Why are these integration tests calling the `http` module? Why aren't we using the CLI to test this stuff? I am interested in the UX. What do we print on success and on failure?

File: dcos/package.py [hunk 4]
```diff
@@ -754,7 +754,7 @@ def update_sources(config, validate=False):
                         source.url,
                         stage_dir)
 
-                    errors.append(e.message)
+                    errors.append(str(e))
```
Comment (jsancio) [inline dcos/package.py]: Why did we change this?

File: dcos/util.py [hunk 5]
```diff
@@ -237,6 +237,22 @@ def configure_process_from_environ():
     configure_debug(os.environ.get(constants.DCOS_DEBUG_ENV))
 
 
+def set_ssl_info_env_vars():
```
Comment (jsancio) [inline dcos/util.py]: Let's move this to `dcoscli.main` no other component should be doing this.

File: dcos/util.py [hunk 6]
```diff
@@ -237,6 +237,22 @@ def configure_process_from_environ():
...
+    """Set SSL info from config to environment variables
+
+    :rtype: None
+    """
+
+    config = get_config()
```
Comment (jsancio) [inline dcos/util.py]: When we move this to `dcosli.main` let's also pass the config in.

File: cli/dcoscli/main.py (hunk 1 above)
Comment (tamarrow) [inline cli/dcoscli/main.py]: Yes, right now we only allow specifying `.pem`.

File: cli/tests/integrations/test_ssl.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    assert resp.status_code == 200
+
+
+def test_verify_ssl():
+    with pytest.raises(DCOSException) as e:
+        http.request("get", DCOS_URL, verify=True)
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Same. Let's test this using the dcos cli. Also, I would like to see the full output from this user story. Ideally in the test. If not possible then a comment in the PR.

File: cli/tests/integrations/test_ssl.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    assert "certificate verify failed" in str(e)
+
+
+def test_verify_ssl_with_bad_cert():
+    with pytest.raises(DCOSException) as e:
+        cert = 'tests/data/ssl/fake.pem'
+        http._request("get", DCOS_URL, verify=True, cert=cert)
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Same. Let's test this using the dcos cli. Also, I would like to see the full output from this user story. Ideally in the test. If not possible then a comment in the PR.

File: cli/tests/integrations/test_ssl.py (hunk 2 above)
Comment (tamarrow) [inline cli/tests/integrations/test_ssl.py]: updated

File: dcos/package.py (hunk 4 above)
Comment (tamarrow) [inline dcos/package.py]: This has always been wrong, and I noticed it when testing.

File: cli/tests/integrations/test_dcos.py [hunk 9]
```diff
@@ -62,6 +62,12 @@ def test_help():
 
     DCOS_DEBUG                  If set then enable further debug messages which
                                 are sent to stdout.
+
+    DCOS_SSL_VERIFY             If set specifies whether to verify SSL certs
+                                for HTTPS.
+
+    DCOS_CERT                   This environment variable points to the
+                                location of the SSL cert .pem file
```
Comment (jsancio) [inline cli/tests/integrations/test_dcos.py]: Let's document this better. Please state that the user can control the value of these environment variables by using `dcos config set ...`.

File: dcos/util.py (hunk 5 above)
Comment (tamarrow) [inline dcos/util.py]: moved

**Comment by jsancio** [comment]
Looks good. Looks like we are missing one test: SSL verification is enabled and we have valid server certificate. Is it possible to test this with the Vagrant DCOS?  cc @cmaloney.

File: dcos/util.py (hunk 6 above)
Comment (tamarrow) [inline dcos/util.py]: ok

File: cli/tests/integrations/test_dcos.py (hunk 9 above)
Comment (tamarrow) [inline cli/tests/integrations/test_dcos.py]: updated

File: cli/dcoscli/main.py [hunk 10]
```diff
@@ -33,6 +33,15 @@
 
     DCOS_DEBUG                  If set then enable further debug messages which
                                 are sent to stdout.
+
+    DCOS_SSL_VERIFY             If set specifies whether to verify SSL certs
+                                for HTTPS. Can also be configured by setting
+                                `core.ssl_config` in the config.
```
Comment (jsancio) [inline cli/dcoscli/main.py]: How about:

```
If set specifies whether to verify SSL certs for HTTPS. To enable permanently use `dcos config set core.ssl_config true`.
```

File: cli/dcoscli/main.py [hunk 11]
```diff
@@ -33,6 +33,15 @@
...
+
+    DCOS_CERT                   This environment variable points to the
+                                location of the SSL cert .pem file. Can also be
+                                configured by setting `core.cert` in the
+                                config.
```
Comment (jsancio) [inline cli/dcoscli/main.py]: How about:

```
This environment variable points to the location of the SSL cert .pem file. To enable permanently
use `dcos config set core.cert <path-to-cert>`.
```

File: cli/dcoscli/main.py (hunk 1 above)
Comment (jsancio) [inline cli/dcoscli/main.py]: Are we going to implement this before merging this PR? If not then let's remove the TODO and create an Issue.

File: cli/dcoscli/main.py [hunk 12]
```diff
@@ -140,3 +154,19 @@ def wrapper(*args, **kwargs):
...
+    """Set SSL info from config to environment variables
+
+    :param config: config
+    :type config: Toml
+    :rtype: None
+    """
+
+    if 'core.ssl_verify' in config:
+        os.environ[constants.DCOS_SSL_VERIFY_ENV] = str(
+            config['core.ssl_verify'])
```
Comment (jsancio) [inline cli/dcoscli/main.py]: This means that `dcos.toml` takes precedence over the `DCOS_SSL_VERIFY` environment variable. Is this what we want? As a user this would be surprising to me.

File: cli/dcoscli/main.py [hunk 13]
```diff
@@ -140,3 +154,19 @@ def wrapper(*args, **kwargs):
...
+
+    if 'core.cert' in config:
+        os.environ[constants.DCOS_CERT_ENV] = config['core.cert']
```
Comment (jsancio) [inline cli/dcoscli/main.py]: This means that dcos.toml takes precedence over the DCOS_CERT environment variable. Is this what we want? As a user this would be surprising to me

File: bin/start_tests.sh [hunk 14]
```diff
@@ -15,6 +15,9 @@ cd cli
 cp tests/data/dcos.toml $DCOS_CONFIG
 sed -i "s/change.dcos.url/$VBOX_IP/g" $DCOS_CONFIG
 
+# overwrite core.dcos_url in the ssl.toml
+sed -i "s/change.dcos.url/$VBOX_IP/g" tests/data/ssl/ssl.toml
```
Comment (jsancio) [inline bin/start_tests.sh]: This means that running the test would change the repository. This is the reason why we `cp` the file in line 15 above.

I think you won't need to do this if you allow environment variable to take precedence over `dcos.toml`. Thoughts?

File: dcos/package.py (hunk 4 above)
Comment (jsancio) [inline dcos/package.py]: :+1:

File: cli/tests/integrations/test_ssl.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+    returncode, stdout, stderr = exec_command(
+        ['dcos', 'marathon', 'app', 'list'], env)
+    assert returncode == 0
+    assert stderr == b''
+
+    config_unset('core.ssl_verify', None, env)
+
+
+def test_verify_ssl_without_ca_env_var(env):
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Technically this should be `test_verify_ssl_withtou_cert_env_var`.

File: cli/tests/integrations/test_ssl.py [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+    returncode, stdout, stderr = exec_command(
+        ['dcos', 'marathon', 'app', 'list'], env)
+    assert returncode == 1
+    assert "certificate verify failed" in stderr.decode('utf-8')
+
+    env.pop(constants.DCOS_SSL_VERIFY_ENV)
+
+
+def test_verify_ssl_without_ca_config(env):
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Technically this should be `test_verify_ssl_withtou_cert_config`.

File: cli/tests/integrations/test_ssl.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
+import os
```
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: Tests look good. What did we decide regarding the self signed cert from the admin router? We are missing a test for valid SSL and HTTPS url.

**Comment by tamarrow** [comment]
@jsancio updated!

**Comment by tamarrow** [comment]
@jsancio all tests pass

File: dcos/http.py [hunk 18]
```diff
@@ -162,11 +174,23 @@ def request(method,
     if 'headers' not in kwargs:
         kwargs['headers'] = {'Accept': 'application/json'}
 
-    response = _request(method, url, is_success, timeout, **kwargs)
+    if verify is None and os.environ.get(constants.DCOS_SSL_VERIFY_ENV):
```
Comment (jsancio) [inline dcos/http.py]: I think you want: `if verify is None and constants.DCOS_SSL_VERIFY_ENV in os.environ:`

File: dcos/http.py [hunk 19]
```diff
@@ -162,11 +174,23 @@ def request(method,
...
+        verify = os.environ.get(constants.DCOS_SSL_VERIFY_ENV)
```
Comment (jsancio) [inline dcos/http.py]: you already checked it about so you want: `os.environ[constants.DCOS_SSL_VERIFY_ENV]`

File: dcos/http.py [hunk 20]
```diff
@@ -162,11 +174,23 @@ def request(method,
...
+        if verify.lower() == "true":
+            verify = True
+        elif verify.lower() == "false":
+            verify = False
+
+    # Silence 'Unverified HTTPS request' and 'SecurityWarning' for bad certs
+    if verify is not None:
+        silence_requests_warnings()
```
Comment (jsancio) [inline dcos/http.py]: I think we do this in a few places. Should we remove them now that we do them here?

File: cli/tests/integrations/test_ssl.py (hunk 17 above)
Comment (jsancio) [inline cli/tests/integrations/test_ssl.py]: @tamarrow Can we test that overriding the CA works? I don't remember if we decided to punt on this.

File: dcos/http.py (hunk 18 above)
Comment (tamarrow) [inline dcos/http.py]: done

File: dcos/http.py (hunk 19 above)
Comment (tamarrow) [inline dcos/http.py]: done

File: dcos/http.py (hunk 20 above)
Comment (tamarrow) [inline dcos/http.py]: I'd rather not, in case some new error surfaces that we are unaware of.

**Comment by tamarrow** [comment]
@jsancio updated

**Comment by jsancio** [comment]
LGTM. Great job @tamarrow !
